src/mpid/ch4/src/ch4r_callbacks.c

```diff
@@ -437,6 +437,7 @@ int MPIDIG_send_target_msg_cb(int handler_id, void *am_hdr, void **data, size_t
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_SEND_TARGET_MSG_CB);
     root_comm = MPIDIG_context_id_to_comm(hdr->context_id);
     if (root_comm) {
+      root_comm_retry:
         /* MPIDI_CS_ENTER(); */
 #ifdef MPIDI_CH4_DIRECT_NETMOD
         rreq = MPIDIG_dequeue_posted(hdr->src_rank, hdr->tag, hdr->context_id,
```

Comment:
@hajimefu I see that you added a lock for the enqueue operation into the unexpected queue. Why do we not need thread safety while dequeuing elements from the posted list?

Reply:
@rzambre It's a little bit non-straightforward, but this lock is actually to protect `MPIDI_global.comm_req_lists[].comm[][]`, not the posted list.

Posted/unexpected request lists are only accessed by netmod/shmmod (and MPIDIG), so they are always protected by a per-vci lock. Given that we currently have only one VCI, there is no concurrency regarding enqueue/dequeue operations to these lists. Of course when we enable multiple VCIs (especially in shmmod, which heavily uses MPIDIG), these lists should be protected as well.

On the other hand, there is a race for `MPIDI_global.comm_req_lists[].comm[][]` even with a single VCI, because in addition to NM/SHM, comm creation/free functions which are outside of per-vci lock would access it.

Reply:
@hajimefu thanks for explaining that, that makes sense. I have some follow up questions but they might be due to my incomplete understanding of the active message logic:

(1) Under what scenarios would `root_coom` equal `NULL` (in line 439, for example)? 
(2) Would it be cleaner to just take the lock before the first `MPIDIG_context_id_to_comm` (line 440, for example) instead of using the current `goto root_comm_retry` logic?

Reply:
@rzambre 

(1) `root_comm` could become NULL when a communicator is being created. If a message exchange (e.g. send-recv) follows immediately after communicator creation, one rank might exit the communicator creation function early and send a message to another rank, while the receiver might have not completed the comm creation. In this case `MPIDIG_context_id_to_comm` would return `NULL` in the receiver side. The tricky thing is that, comm creation might have finished while the receiver is running `MPIDIG_send_target_msg_cb` (remember we are talking about the handoff model, where the user thread executes comm creation logic and the progress thread executes the callback function). This is why I had to check `root_comm_again` later.

**Without** this check, what happens is that
a. (progress thread) A message arrives and `MPIDIG_send_target_msg_cb` is called, and `root_comm` is `NULL` (L444)
b. (user thread) `MPIDIG_init_comm` is executed and it moves all unexpected messages arrived early to the comm (L54-63)
c. (progress thread) `MPIDIG_send_target_msg_cb` inserts the message into the early-arrived unexpected list (L525), which is already processed by the step above and therefore never be processed in the future. This message never arrives to the user program and causes deadlock.

(2) It might work (though I haven't verified yet). I just wanted to minimize the lock duration.

Reply:
Thank you for the explanation with example. Everything else looks good to me.